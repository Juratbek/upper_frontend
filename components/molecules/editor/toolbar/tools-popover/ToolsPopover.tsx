import { CSSProperties, forwardRef } from 'react';

import { useEditorContext } from '../../context/useEditorContext';
import { TToolType } from '../../tools/tool.types';
import { Item, Popover } from '../popover/Popover';
import classes from './ToolsPopover.module.scss';

export const ToolsPopover = forwardRef<
  HTMLDivElement,
  { open: boolean; close: VoidFunction; style: CSSProperties }
>(function Component({ open, close, style }, ref) {
  const { tools, addBlock, hoveredBlock } = useEditorContext();

  const itemClickHandler = (type: TToolType) => {
    if (!hoveredBlock) {
      console.error("Tooltip shouldn't be shown if there is no hovered block");
      return;
    }
    addBlock(type, hoveredBlock.id);
    close();
  };

  const toolTypes = Object.keys(tools) as TToolType[];

  return (
    <Popover open={open} ref={ref} style={style}>
      <div className={classes.list}>
        {toolTypes.map((type) => {
          const { toolbar } = tools[type];
          if (!toolbar) return null;

          return (
            <Item onClick={() => itemClickHandler(type)} key={toolbar.text} icon={toolbar.icon}>
              {toolbar.text}
            </Item>
          );
        })}
      </div>
    </Popover>
  );
});
