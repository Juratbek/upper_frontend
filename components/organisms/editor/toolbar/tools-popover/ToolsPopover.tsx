import { CSSProperties, forwardRef } from 'react';
import { detectPlatform } from 'utils';

import { useEditorContext } from '../../context/useEditorContext';
import { TShortcut, TToolType } from '../../tools/tool.types';
import { Item, Popover } from '../popover/Popover';
import classes from './ToolsPopover.module.scss';

export const ToolsPopover = forwardRef<
  HTMLDivElement,
  { open: boolean; close: VoidFunction; style: CSSProperties }
>(function Component({ open, close, style }, ref) {
  const { tools, addBlock, hoveredBlock } = useEditorContext();

  const itemClickHandler = (type: TToolType) => {
    if (!hoveredBlock) {
      console.error("Tooltip shouldn't be shown if there is no hovered block");
      return;
    }
    addBlock(type, hoveredBlock.id);
    close();
  };

  const toolTypes = Object.keys(tools) as TToolType[];

  return (
    <Popover open={open} ref={ref} style={style}>
      <div className={classes.list}>
        {toolTypes.map((type) => {
          const { toolbar, shortcuts } = tools[type];
          if (!toolbar) return null;

          return (
            <Item
              onClick={() => itemClickHandler(type)}
              key={toolbar.text}
              icon={toolbar.icon}
              suffix={Array.isArray(shortcuts) ? shortcuts.map(getShortcutLabel).join(', ') : null}
            >
              {toolbar.text}
            </Item>
          );
        })}
      </div>
    </Popover>
  );
});

function getShortcutLabel(shortcut: TShortcut): string {
  let shortcutStr = shortcut;
  if (typeof shortcut === 'object') {
    shortcutStr = shortcut.key;
  }

  const platform = detectPlatform();
  if (platform === 'MacOS') return `CMD+${shortcutStr}`;
  if (platform === 'Linux' || platform === 'Windows') return `CTRL+${shortcutStr}`;

  return '';
}
