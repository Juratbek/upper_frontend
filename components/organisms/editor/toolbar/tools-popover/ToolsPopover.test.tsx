import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CSSProperties } from 'react';

import { ToolsPopover } from './ToolsPopover';

const mocks = vi.hoisted(() => ({
  addBlock: vi.fn(),
}));

vi.mock('../../context/useEditorContext', async (importOriginal) => {
  const original = await importOriginal<typeof import('../../context/useEditorContext')>();

  return {
    ...original,
    useEditorContext: () => ({
      ...original.useEditorContext(),
      addBlock: mocks.addBlock,
      hoveredBlock: {
        id: 'unique_id',
      },
    }),
  };
});

vi.mock('utils', async (importOriginal) => {
  const original = await importOriginal<typeof import('utils')>();
  return {
    ...original,
    detectPlatform: vi.fn().mockReturnValue('Linux'),
  };
});

const mockProps: { open: boolean; close: VoidFunction; style: CSSProperties } = {
  open: true,
  close: vi.fn(),
  style: {},
};

describe('ToolsPopover', () => {
  it('calls addBlock and close on item click', async () => {
    const mockClose = vi.fn();
    render(<ToolsPopover {...mockProps} close={mockClose} />);
    const button = screen.getByRole('button', { name: /Matn/i });

    userEvent.click(button);

    await waitFor(() => {
      expect(mocks.addBlock).toHaveBeenCalledTimes(1);
      expect(mockClose).toHaveBeenCalled();
    });
  });

  it('snapshot', () => {
    render(<ToolsPopover {...mockProps} />);
    expect(document.body).toMatchSnapshot();
  });
});
