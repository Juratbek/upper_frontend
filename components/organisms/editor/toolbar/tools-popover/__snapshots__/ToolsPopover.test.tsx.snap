// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToolsPopover > snapshot 1`] = `
<body>
  <div>
    <div
      class="popover open"
      role="dialog"
    >
      <div
        class="list"
      >
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 9V7.2C8 7.08954 8.08954 7 8.2 7L12 7M16 9V7.2C16 7.08954 15.9105 7 15.8 7L12 7M12 7L12 17M12 17H10M12 17H14"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Matn
          </div>
          <span
            class="suffix"
          >
            CTRL+m
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 7L9 12M9 17V12M9 12L15 12M15 7V12M15 17L15 12"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Sarlavha
          </div>
          <span
            class="suffix"
          >
            CTRL+1, CTRL+2, CTRL+3, CTRL+4, CTRL+5, CTRL+6
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="14"
                rx="4"
                stroke="currentColor"
                stroke-width="2"
                width="14"
                x="5"
                y="5"
              />
              <path
                d="M5.13968 15.32L8.69058 11.5661C9.02934 11.2036 9.48873 11 9.96774 11C10.4467 11 10.9061 11.2036 11.2449 11.5661L15.3871 16M13.5806 14.0664L15.0132 12.533C15.3519 12.1705 15.8113 11.9668 16.2903 11.9668C16.7693 11.9668 17.2287 12.1705 17.5675 12.533L18.841 13.9634"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
              <path
                d="M13.7778 9.33331H13.7867"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Rasm
          </div>
          <span
            class="suffix"
          >
            CTRL+r
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 8L5 12L9 16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
              <path
                d="M15 8L19 12L15 16"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Kod
          </div>
          <span
            class="suffix"
          >
            CTRL+k
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
                x1="6"
                x2="10"
                y1="12"
                y2="12"
              />
              <line
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
                x1="14"
                x2="18"
                y1="12"
                y2="12"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Ajratuvchi
          </div>
          <span
            class="suffix"
          >
            CTRL+/
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
                x1="9"
                x2="19"
                y1="7"
                y2="7"
              />
              <line
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
                x1="9"
                x2="19"
                y1="12"
                y2="12"
              />
              <line
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
                x1="9"
                x2="19"
                y1="17"
                y2="17"
              />
              <path
                d="M5.00001 17H4.99002"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
              <path
                d="M5.00001 12H4.99002"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
              <path
                d="M5.00001 7H4.99002"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Ro'yxat
          </div>
          <span
            class="suffix"
          >
            CTRL+l
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="14"
                rx="4"
                stroke="currentColor"
                stroke-width="2"
                width="14"
                x="5"
                y="5"
              />
              <line
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
                x1="12"
                x2="12"
                y1="9"
                y2="12"
              />
              <path
                d="M12 15.02V15.01"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Ogohlantiruvchi
          </div>
          <span
            class="suffix"
          >
            CTRL+o
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 10.8182L9 10.8182C8.80222 10.8182 8.60888 10.7649 8.44443 10.665C8.27998 10.5651 8.15181 10.4231 8.07612 10.257C8.00043 10.0909 7.98063 9.90808 8.01922 9.73174C8.0578 9.55539 8.15304 9.39341 8.29289 9.26627C8.43275 9.13913 8.61093 9.05255 8.80491 9.01747C8.99889 8.98239 9.19996 9.00039 9.38268 9.0692C9.56541 9.13801 9.72159 9.25453 9.83147 9.40403C9.94135 9.55353 10 9.72929 10 9.90909L10 12.1818C10 12.664 9.78929 13.1265 9.41421 13.4675C9.03914 13.8084 8.53043 14 8 14"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
              <path
                d="M16 10.8182L15 10.8182C14.8022 10.8182 14.6089 10.7649 14.4444 10.665C14.28 10.5651 14.1518 10.4231 14.0761 10.257C14.0004 10.0909 13.9806 9.90808 14.0192 9.73174C14.0578 9.55539 14.153 9.39341 14.2929 9.26627C14.4327 9.13913 14.6109 9.05255 14.8049 9.01747C14.9989 8.98239 15.2 9.00039 15.3827 9.0692C15.5654 9.13801 15.7216 9.25453 15.8315 9.40403C15.9414 9.55353 16 9.72929 16 9.90909L16 12.1818C16 12.664 15.7893 13.1265 15.4142 13.4675C15.0391 13.8084 14.5304 14 14 14"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Iqtibos
          </div>
          <span
            class="suffix"
          >
            CTRL+i
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 5V18.5"
                stroke="currentColor"
                stroke-width="2"
              />
              <path
                d="M5 10H19"
                stroke="currentColor"
                stroke-width="2"
              />
              <rect
                height="14"
                rx="4"
                stroke="currentColor"
                stroke-width="2"
                width="14"
                x="5"
                y="5"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            Jadval
          </div>
          <span
            class="suffix"
          >
            CTRL+j
          </span>
        </button>
        <button
          class="btn"
        >
          <span
            class="icon"
          >
            <svg
              fill="none"
              height="24"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                height="14"
                rx="4"
                stroke="currentColor"
                stroke-width="2"
                width="14"
                x="5"
                y="5"
              />
              <path
                d="M12 15.52V15.51"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
              <path
                d="M10.0024 9.97655C10.1567 9.01858 11 8.5 12 8.5C13 8.5 13.6857 9.17188 13.8693 9.70703C14.0529 10.2422 14.0135 11.0514 13.5067 11.5159C13 11.9805 12.7344 11.832 12.2784 12.3168C12.1134 12.4923 12 12.7476 12 12.7476"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              />
            </svg>
          </span>
          <div
            class="text"
          >
            So'rovnoma
          </div>
          <span
            class="suffix"
          >
            CTRL+s
          </span>
        </button>
      </div>
    </div>
  </div>
</body>
`;
